File: HW7/collcount.fs
( File:   collcount.fs                )
( Author: Andrew Player, 2022         )
( ----------------------------------- )
( Counts the stopping time for the    )
( Collatz Conjecture for a given num. )

( Collatz Operation )
( ----------------- )
: coll 
    dup 
    1 = 
        if
        else 
            dup 
            2 mod 0 = 
                if  
                    2 / 
                else 
                    3 * 1 + 
                then 
                    swap 1 + swap 
        then ;

( Collatz Loop )
( ------------ )
: collcount 
    0 swap 
    begin 
        dup 
        1 > 
        while 
            coll 
        repeat 
            drop ;

File: HW7/collcount.pl
% Filename: collcount.pl
% Author:   Andrew Player
% Year:     2022 

% Queries if C is the stopping time for
% N in the Collatz Conjecture
collcount(N, C) :- 
(
    N =:= 1 -> 
        C =:= 0 
        ; 
        (N mod 2 =:= 0 -> 
            collcount(N div 2, C - 1) 
            ; 
            collcount(3 * N + 1, C - 1)
        )
).